server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java

```diff
@@ -173,6 +175,103 @@ public void testClusterRecovered() {
+            .build();
+        final var clusterState1 = ClusterState.builder(clusterState0)
+            .metadata(metadata1)
+            .routingTable(RoutingTable.builder(clusterState0.routingTable()).addAsFromOpenToClose(metadata1.index("test")))
+            .build();
+
+        assertLastAllocatedNodeIdsAssigned(
+            UnassignedInfo.Reason.INDEX_CLOSED,
+            clusterState0.routingTable().index("test"),
+            clusterState1.routingTable().index("test")
+        );
+
+        // cluster state 2: closed index has been fully assigned
+        final var clusterState2 = applyStartedShardsUntilNoChange(clusterState1, allocationService);
+        assertTrue(clusterState2.routingTable().index("test").allShardsActive());
+
+        // cluster state 3: index reopened, fully unassigned again
+        final var metadata3 = Metadata.builder(metadata0)
+            .put(IndexMetadata.builder(metadata1.index("test")).state(IndexMetadata.State.OPEN))
+            .build();
+        final var clusterState3 = ClusterState.builder(clusterState2)
+            .metadata(metadata3)
+            .routingTable(RoutingTable.builder(clusterState2.routingTable()).addAsFromCloseToOpen(metadata3.index("test")))
+            .build();
+
+        assertLastAllocatedNodeIdsAssigned(
+            UnassignedInfo.Reason.INDEX_REOPENED,
+            clusterState2.routingTable().index("test"),
+            clusterState3.routingTable().index("test")
+        );
+    }
+
+    private void assertLastAllocatedNodeIdsAssigned(
+        UnassignedInfo.Reason expectedUnassignedReason,
+        IndexRoutingTable originalRoutingTable,
+        IndexRoutingTable finalRoutingTable
+    ) {
+        boolean foundAnyNodeIds = false;
+        for (int shardId = 0; shardId < finalRoutingTable.size(); shardId++) {
+            final var previousShardRoutingTable = originalRoutingTable.shard(shardId);
+            final var previousNodes = previousShardRoutingTable == null
+                ? Set.<String>of()
+                : IntStream.range(0, previousShardRoutingTable.size()).mapToObj(previousShardRoutingTable::shard).map(shard -> {
+                    assertTrue(shard.started());
+                    return shard.currentNodeId();
+                }).collect(Collectors.toSet());
+            final var shardRoutingTable = finalRoutingTable.shard(shardId);
+            for (int shardCopy = 0; shardCopy < shardRoutingTable.size(); shardCopy++) {
+                final var shard = shardRoutingTable.shard(shardCopy);
+                assertTrue(shard.unassigned());
+                assertThat(shard.unassignedInfo().getReason(), equalTo(expectedUnassignedReason));
+                final var lastAllocatedNodeId = shard.unassignedInfo().getLastAllocatedNodeId();
+                if (lastAllocatedNodeId == null) {
+                    // restoring an index may change the number of shards/replicas so no guarantee that lastAllocatedNodeId is populated
```

Comment:
I think only the number of replicas, not the number of shards can be changed? Probably what you meant with shards/replicas, but removing "shards/" would be better I think.
```suggestion
                    // restoring an index may change the number of replicas so no guarantee that lastAllocatedNodeId is populated
```

Reply:
On the contrary, I didn't think there's anything to require that the snapshot has the same number of shards as index on top of which it's being restored.

Reply:
Ahh, right, thanks.